bson/json_util.py

```diff
@@ -855,100 +855,189 @@ def _encode_int64(obj: Int64, json_options: JSONOptions) -> Any:
-        return obj
-    elif isinstance(obj, float):
-        if json_options.json_mode != JSONMode.LEGACY:
-            if math.isnan(obj):
-                return {"$numberDouble": "NaN"}
-            elif math.isinf(obj):
-                representation = "Infinity" if obj > 0 else "-Infinity"
-                return {"$numberDouble": representation}
-            elif json_options.json_mode == JSONMode.CANONICAL:
-                # repr() will return the shortest string guaranteed to produce the
-                # original value, when float() is called on it.
-                return {"$numberDouble": str(repr(obj))}
-        return obj
-    elif isinstance(obj, str):
-        return obj
-    elif isinstance(obj, datetime.datetime):
-        if json_options.datetime_representation == DatetimeRepresentation.ISO8601:
-            if not obj.tzinfo:
-                obj = obj.replace(tzinfo=utc)
-                assert obj.tzinfo is not None
-            if obj >= EPOCH_AWARE:
-                off = obj.tzinfo.utcoffset(obj)
-                if (off.days, off.seconds, off.microseconds) == (0, 0, 0):  # type: ignore
-                    tz_string = "Z"
-                else:
-                    tz_string = obj.strftime("%z")
-                millis = int(obj.microsecond / 1000)
-                fracsecs = ".%03d" % (millis,) if millis else ""
-                return {
-                    "$date": "{}{}{}".format(obj.strftime("%Y-%m-%dT%H:%M:%S"), fracsecs, tz_string)
-                }
-
-        millis = _datetime_to_millis(obj)
-        if json_options.datetime_representation == DatetimeRepresentation.LEGACY:
-            return {"$date": millis}
-        return {"$date": {"$numberLong": str(millis)}}
-    elif isinstance(obj, bytes):
-        return _encode_binary(obj, 0, json_options)
-    elif isinstance(obj, uuid.UUID):
-        if json_options.strict_uuid:
-            binval = Binary.from_uuid(obj, uuid_representation=json_options.uuid_representation)
-            return _encode_binary(binval, binval.subtype, json_options)
-        else:
-            return {"$uuid": obj.hex}
+    # First see if the type is already cached. KeyError will only ever
+    # happen once per subtype.
+    obj_type = type(obj)
+    try:
+        return _ENCODERS[obj_type](obj, json_options)
+    except KeyError:
+        pass
+
+    # Second, fall back to trying _type_marker. This has to be done
+    # before the loop below since users could subclass one of our
+    # custom types that subclasses a python built-in (e.g. Binary)
+    if hasattr(obj, "_type_marker"):
+        marker = obj._type_marker
+        if marker in _MARKERS:
+            func = _MARKERS[marker]
+            # Cache this type for faster subsequent lookup.
+            _ENCODERS[obj_type] = func
+            return func(obj, json_options)
+
+    # Third, test each base type. This will only happen once for
+    # a subtype of a supported base type.
+    for base in _BUILT_IN_TYPES:
```

Comment:
And this is the same for non-BSON types?

Reply:
This is to handle cases like:
```python
        class myint(int):
            pass
```

Reply:
Makes sense, I think the comment would be clearer if it explicitly said why this is needed.